lib/rust/profiler/src/lib.rs

```diff
@@ -0,0 +1,333 @@
+            pub static NEXT_ID: cell::Cell<u32> = cell::Cell::new(1);
+        }
+        ProfilerId(NEXT_ID.with(|next_id| {
+            let id = next_id.get();
+            next_id.set(id + 1);
+            id
+        }))
+    }
+}
+
+/// Pseudo-profiler serving as the root of the measurement hierarchy.
+pub const APP_LIFETIME: Objective = Objective(ProfilerId(0));
+
+
+
+// =======================
+// === StartedProfiler ===
+// =======================
+
+/// The interface supported by profiler-data objects.
+pub trait StartedProfiler {
+    fn finish(&self, profiler: ProfilerId);
+}
+
+// === Implementation for disabled profilers ===
+
+impl StartedProfiler for () {
+    fn finish(&self, _: ProfilerId) {}
+}
+
+// ====================
+// === ProfilerData ===
+// ====================
+
+/// Data used by a started Measurement for an enabled profile level.
+pub struct ProfilerData {
+    pub parent: ProfilerId,
+    pub start:  Option<Timestamp>,
+    pub label:  Label,
+}
+
+// === Trait Implementations ===
+
+impl StartedProfiler for ProfilerData {
+    fn finish(&self, profiler: ProfilerId) {
+        let parent = self.parent;
+        let start = self.start;
+        let label = self.label;
+        let end = Timestamp::now();
+        let measurement = Measurement { parent, profiler, start, end, label };
+        crate::MEASUREMENTS.with(move |log| log.push(measurement));
+    }
+}
+
+
+
+// ===================
+// === Measurement ===
+// ===================
+
+/// Identifies a profiled section, the parent it was reached by, and its entry and exit times.
+pub struct Measurement {
+    pub parent:   ProfilerId,
+    pub profiler: ProfilerId,
+    pub start:    Option<Timestamp>,
+    pub end:      Timestamp,
+    pub label:    Label,
+}
+
+pub mod internal {
+    use crate::*;
+
+    use std::cell;
+    use std::mem;
+
+    // =======================
+    // === LocalVecBuilder ===
+    // =======================
+
+    /// Data structure supporting limited interior mutability, to build up a collection.
+    pub struct LocalVecBuilder<T>(cell::UnsafeCell<Vec<T>>);
+    impl<T> LocalVecBuilder<T> {
+        #[allow(clippy::new_without_default)]
+        /// Create a new, empty vec builder.
+        pub fn new() -> Self {
+            Self(cell::UnsafeCell::new(vec![]))
+        }
+
+        /// Push an element.
+        pub fn push(&self, element: T) {
+            // Note [LocalVecBuilder Safety]
+            unsafe {
+                (&mut *self.0.get()).push(element);
+            }
+        }
+
+        // Note [LocalVecBuilder Safety]
+        // =============================
+        // When obtaining a reference from the UnsafeCell, all accessors follow these rules:
+        // - There must be a scope that the reference doesn't escape.
+        // - There must be no other references obtained in the same scope.
+        // Consistently following these rules ensures the no-alias rule of mutable references is
+        // satisfied.
```

Comment:
There is one function referring this note below the note and one above the note. Move the note below both of them for just a tiny better clarity when reading the code, please :)